File: c/src/day5.c
#include "util.c"
#include <limits.h>
#include <ctype.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

#define MAPS 7

typedef struct Range {
	long from;
	long to;
} Range;

bool range_contains(Range* r, long i) {
	return r->from <= i && i <= r->to;
}

typedef struct {
	Range src_range;
	Range dest_range;
} MapEntry;

#define TYPE MapEntry
#define TYPED(t) MapEntry ## t
#include "darray.h"

size_t parse_nums(long* out, char* line, size_t len) {
	size_t sc = 0;
	for (size_t i = 0; i < len; i++) {
		char curr = line[i];
		if (isdigit(curr)) {
			long n = 0;
			while(isdigit(line[i])) {
				n = concat(n, line[i]);
				i++;
			}
			out[sc++] = n;
		}
	}
	return sc;
}

long day5_part1() {
	FILE* f = open_file("../../input/5.txt");

	char* line = (char*) malloc(sizeof(char)*LINE_LEN);
	ssize_t len = 0;

	long seeds[32] = {0};
	MapEntryArr arr[MAPS] = { 0 };
	for (size_t i = 0; i < MAPS; i++) {
		arr[i] = MapEntryArr_new();
	}

	len = read_line(f, LINE_LEN, &line);
	size_t seedc = parse_nums(seeds, line, len);
	size_t arrc = 0;
	bool scanning = false;
	while ((len = read_line(f, LINE_LEN, &line)) != -1) {
		if (isdigit(line[0])) {
			scanning = true;
			long map[3] = { 0 };
			assert(parse_nums(map, line, len) == 3);
			Range dstrng = { .from = map[0], .to = map[0] + map[2] };
			Range srcrng = { .from = map[1], .to = map[1] + map[2] };
			MapEntry me = { .src_range = srcrng, .dest_range = dstrng };
			MapEntryArr_add(&arr[arrc], me);
		} else if (scanning) {
			scanning = false;
			arrc++;
		}
	}

	long min = LONG_MAX;
	for (size_t i = 0; i < seedc; i++) {
		long s = seeds[i];
		for (size_t j = 0; j < MAPS; j++) {
			for (size_t k = 0; k < arr[j].len; k++) {
				MapEntry me = MapEntryArr_get(&arr[j], k);
				if (range_contains(&me.src_range, s)) {
					s = me.dest_range.from + s - me.src_range.from;
					assert(s > 0);
					break;
				}
			}
		}
		min = min(min, s);
	}

	for (size_t i = 0; i < MAPS; i++) {
		MapEntryArr_free(&arr[i]);
	}

	free(line);
	fclose(f);

	return min;
}

long day5_part2() {
	return 0;
}

File: go/day5.go
package main

import (
	"bufio"
	"math"
	"os"
	"sync"
)

type MapEntry struct {
	Src Range
	Dst Range
}

func parse_nums(line []byte) []int64 {
	var s []int64
	i := 0
	for i < len(line) {
		if isDigit(line[i]) {
			var n int64 = 0
			for i < len(line) && isDigit(line[i]) {
				n = concat(n, line[i])
				i++
			}
			s = append(s, n)
		} else {
			i++
		}
	}
	return s

}

func computeRanges(seeds []int64) []Range {
	var r []Range
	for i := 0; i < len(seeds); i += 2 {
		r = append(r, Range{From: seeds[i], To: seeds[i] + seeds[i+1]})
	}
	return r
}

func processRange(rng Range, ch chan int64, wg *sync.WaitGroup, maps [7][]MapEntry) {
	var res int64 = math.MaxInt64
	for s := rng.From; s <= rng.To; s++ {
		ss := s
		for _, m := range maps {
			for _, entry := range m {
				if entry.Src.Contains(ss) {
					ss = entry.Dst.From + ss - entry.Src.From
					break
				}
			}
		}
		res = min(res, ss)
	}
	ch <- res
	wg.Done()
}

func Day5Part2() int64 {
	f, _ := os.OpenFile("../input/5.txt", os.O_RDONLY, 0)
	defer f.Close()
	scanner := bufio.NewScanner(f)
	scanner.Scan()
	seeds := parse_nums(scanner.Bytes())
	var maps [7][]MapEntry
	mapc := 0
	scanning := false
	for scanner.Scan() {
		line := scanner.Bytes()
		if len(line) > 0 && isDigit(line[0]) {
			scanning = true
			nums := parse_nums(line)
			assert(len(nums) == 3)
			dstrng := Range{From: nums[0], To: nums[0] + nums[2]}
			srcrng := Range{From: nums[1], To: nums[1] + nums[2]}
			entry := MapEntry{Src: srcrng, Dst: dstrng}
			maps[mapc] = append(maps[mapc], entry)
		} else if scanning {
			mapc += 1
			scanning = false
		}
	}

	seedRanges := computeRanges(seeds)

	var wg sync.WaitGroup
	ch := make(chan int64)
	for _, rng := range seedRanges {
		wg.Add(1)
		go processRange(rng, ch, &wg, maps)
	}

	go func() {
		wg.Wait()
		close(ch)
	}()

	var result int64 = math.MaxInt64
	for r := range ch {
		result = min(result, r)
	}

	return result
}
